### DISCUSSION TITLE
LiveMap

### DISCUSSION THREAD

**Discussion by cixio** [dl9rdz/rdz_ttgo_sonde#126] [TTGO firmware]
I started creating a live map in the web view a little over a year ago. On this one can see the current sonde position on a map as well as the previous (received) course and the current prediction.

The map works on all modern browsers (iOS / Android / PC / Mac) if there is an internet connection.

There are a few changes in the code, overall we are only talking about about 15kb of additional storage space that is required since most things are integrated externally (map data via wikimedia, jQuery, leaflet and prediction via tawhiri). I'm not sure about GDPR with regard to the external integrations...?

I've attached a couple of screenshots. Is there a general interest in this extension? I can create a pull request if desired.

![rdzTTGOSonde Server LiveMap 1](https://user-images.githubusercontent.com/5869764/128704328-732612af-c7ad-4c39-bf23-49c95a9ec5e6.png)
![rdzTTGOSonde Server LiveMap 2](https://user-images.githubusercontent.com/5869764/128704345-7fc78e70-46cf-4d1a-b1d9-72319240bb01.png)
![rdzTTGOSonde Server LiveMap 3](https://user-images.githubusercontent.com/5869764/128704353-ade44ea0-74bc-49d8-885f-e37677798e0f.png)

**Comment by cixio** [comment]
Adittional comment: if you use a TTGO with GPS your own position is also displayed in the map.

**Comment by 5B4APV** [comment]
Looks very nice, it's a useful add-on in my opinion. I would like to ask if it's possible to store the map on the sd card of the TTGO and have an offline mode.

**Reply by cixio** [reply]
Not planned. Also as far as I know the SD card reader is not in use on rdzTTGOsonde?

**Reply by LukePrior** [reply]
> Not planned. Also as far as I know the SD card reader is not in use on rdzTTGOsonde?

How much space would adding support take as it would be great for offline usage albeit without predictions.

**Reply by dl9rdz** [reply]
Its not so much a problem of code space (well, maybe it is, the SD card file system code might take some space), but more of a RAM issue. The ESP32 web server starts struggeling randomly if you have a single data item of about >10k size to be transferred. I am quite worried that using the ESP32 as something like a tile server will render everything highly unstable.

Also I have not found an easy way for connecting the leaflet map to an offline map, otherwise I would support offline maps in the android app. Offline prediction would be feasible (an an Android phone, not on the TTGO) if you prefetch some weather data.

**Reply by LukePrior** [reply]
> Its not so much a problem of code space (well, maybe it is, the SD card file system code might take some space), but more of a RAM issue. The ESP32 web server starts struggeling randomly if you have a single data item of about >10k size to be transferred. I am quite worried that using the ESP32 as something like a tile server will render everything highly unstable.

Yes I could see that happening probably best to leave it be :)

> Also I have not found an easy way for connecting the leaflet map to an offline map, otherwise I would support offline maps in the android app. 

This would be a **much** better solution and I do believe a few options exist. The two main ones I can find are listed below:

- https://github.com/gregallensworth/L.TileLayer.Cordova
- https://github.com/MazeMap/Leaflet.TileLayer.PouchDBCached

> Offline prediction would be feasible (on an Android phone, not on the TTGO) if you prefetch some weather data.

This would be a great addition to the app I believe Chasemapper has a similar feature for offline predictions using the [CUSF Standalone Predictor:](https://github.com/darksidelemm/cusf_predictor_wrapper) however I'm not sure if that could be implemented on mobile?

**Comment by dl9rdz** [comment]
Looks nice. 15k seems a load (on pio builds only 25k are still available), is that all in the code part? Maybe parts can be off-loaded to the file system (which has more space available).  I will have a closer look if you send a pull request.

**Reply by cixio** [reply]
~15kb on the file system for additions in css, js file and html file

on the code part:
- added accessing files on the file system
- created a json export for the data
- changed position of GPS buffer(?) to get the GPS data in the json part (you will see what I mean in the pull request)

Maybe the code part can be optimized as I am not familiar with the language it is written in (git says c++? not sure).

**Comment by LukePrior** [comment]
If you wanted you could add the custom icons we use for the SondeHub tracker to the map:

https://github.com/projecthorus/sondehub-tracker/tree/master/img/markers
